File: 101/scala/src/main/scala/BinarySearch.scala
/**
  * Perform a binary search.
  */
object BinarySearch {

  /**
    * Perform a binary search for an Int.
    *
    * @param seq the seq to search.
    * @param term the term to search for.
    * @return true if found, false if not.
    */
  def apply(seq: Seq[Int], term: Int): Boolean = {
    var low = 0
    var hi = seq.size - 1

    while (low <= hi) {
      val mid = (low + hi) / 2

      if (term > seq(mid)) {
        low = mid + 1
      } else if (term < seq(mid)) {
        hi = mid - 1
      } else {
        return true
      }
    }
    false
  }
}

File: 101/ts/src/binarySearch.ts
/**
 * Perform a binary search for an Int.
 *
 * @param {Array} arr the array to search.
 * @param {Number} term the term to search for.
 * @return {Boolean} true if found, false if not.
 */
export const binarySearch = (arr: Array<number>, term: number): boolean => {
  let lo = 0;
  let hi = arr.length - 1;

  while (hi > lo) {
    const mid = (lo + hi) / 2;
    const value = arr[mid];

    if (value === term) {
      return true;
    } else if (value > term) {
      hi = mid - 1;
    } else {
      lo = mid + 1;
    }
  }

  return false;
};
